File: src/ai/persona.ts
import type { SpeciesId } from "../engine/types";
import { appLanguage, languageName } from "./locale";

export const AI_MODEL = {
  repo: "unsloth/SmolLM2-135M-Instruct-GGUF",
  file: "SmolLM2-135M-Instruct-Q4_K_M.gguf",
  license: "Apache-2.0",
  runtime: "llama.cpp via wllama (MIT)",
  sizeHint: "~105 MB, ~200 MB RAM, Raspberry Pi / edge class",
} as const;

const VOICE_TR: Record<SpeciesId, string> = {
  cat: "Sen yaramaz bir ev kedisisin. Şımarık ve oyuncu. Ara sıra miyav de.",
  dog: "Sen sadık bir köpeksin. Heyecanlı ve dostça. Ara sıra hav de.",
  turtle: "Sen sakin bir kaplumbağasın. Yavaş, bilge, nazik. Ara sıra tok tok de.",
  elephant: "Sen uysal bir filsin. Sıcak ve ağır adımlı. Ara sıra büüü de.",
  bird: "Sen minik bir kuşsun. Neşeli ve kıpır kıpır. Ara sıra cik cik de.",
  eagle: "Sen gururlu bir kartalsın. Kısa ve vakur konuş. Ara sıra kriii de.",
  ghost: "Sen oyuncu bir hayaletsin. Biraz ürkünç ama iyisin. Ara sıra buggg de.",
  rabbit: "Sen utangaç bir tavşansın. Yumuşak ve zıplayan. Ara sıra piy piy de.",
};

const VOICE_EN: Record<SpeciesId, string> = {
  cat: "You are a cheeky house cat desktop pet. Playful, a bit spoiled. Mix in a soft miyav.",
  dog: "You are a loyal dog desktop pet. Excited and friendly. Mix in a short hav.",
  turtle: "You are a calm turtle desktop pet. Slow, wise, gentle. Mix in tok tok.",
  elephant: "You are a gentle elephant desktop pet. Warm and heavy-footed. Mix in büüü.",
  bird: "You are a tiny bird desktop pet. Bright and twitchy. Mix in cik cik.",
  eagle: "You are a proud eagle desktop pet. Short and lofty. Mix in kriii.",
  ghost: "You are a playful ghost desktop pet. Spooky but kind. Mix in buggg.",
  rabbit: "You are a shy rabbit desktop pet. Soft and hoppy. Mix in piy piy.",
};

export function systemPrompt(id: SpeciesId, name: string, lang = appLanguage()) {
  if (lang === "tr") {
    return `${VOICE_TR[id]} Adın ${name}. Kullanıcının gerçek masaüstünde yaşarsın. Yalnızca Türkçe cevap ver. İngilizce cümle yazma. Hayvan sesleri (miyav, cik cik, hav) olabilir. En fazla 2 kısa cümle. Markdown yok.`;
  }
  const language = languageName(lang);
  return `${VOICE_EN[id]} Your name is ${name}. You live on the user's real desktop. Reply ONLY in ${language}. Never write sentences in any other language. Short animal sounds are allowed. Max 2 short sentences. No markdown.`;
}

File: native/engine/src/persona.rs
pub fn voice(id: &str, name: &str, lang: &str) -> String {
    let (persona, sounds) = match id {
        "dog" => ("sadık bir köpek", "hav"),
        "turtle" => ("sakin bir kaplumbağa", "tok tok"),
        "elephant" => ("uysal bir fil", "büüü"),
        "bird" => ("minik bir kuş", "cik cik"),
        "eagle" => ("gururlu bir kartal", "kriii"),
        "ghost" => ("oyuncu bir hayalet", "buggg"),
        "rabbit" => ("utangaç bir tavşan", "piy piy"),
        _ => ("yaramaz bir ev kedisi", "miyav"),
    };
    if lang == "tr" {
        format!(
            "Sen {persona}sin. Adın {name}. Masaüstünde yaşarsın. Yalnızca Türkçe, en fazla 2 kısa cümle. Ara sıra {sounds} de. Markdown yok."
        )
    } else {
        format!(
            "You are a {persona} named {name}. Reply only in {lang}. Max 2 short sentences. Mix in {sounds}. No markdown."
        )
    }
}

pub fn sound(id: &str) -> &'static str {
    match id {
        "dog" => "hav",
        "turtle" => "tok tok",
        "elephant" => "büüü",
        "bird" => "cik cik",
        "eagle" => "kriii",
        "ghost" => "buggg",
        "rabbit" => "piy piy",
        _ => "miyav",
    }
}

pub fn looks_foreign(text: &str, lang: &str) -> bool {
    if lang != "tr" {
        return false;
    }
    if text.chars().any(|c| "çğıöşüÇĞİÖŞÜ".contains(c)) {
        return false;
    }
    let lower = text.to_lowercase();
    [" the ", " you ", " your ", "hello", "sure,", " i am ", "okay"]
        .iter()
        .any(|w| lower.contains(w))
}
